qiskit/transpiler/passes/synthesis/high_level_synthesis.py

```diff
@@ -757,7 +823,29 @@ def run(self, high_level_object, coupling_map=None, target=None, qubits=None, **
         if use_inverted:
             mat = calc_inverse_matrix(mat)
 
-        decomposition = synth_cnot_depth_line_kms(mat)
+        decomposition = synth_cnot_depth_line_kms(mat, use_dag=use_dag)
+
+        if use_dag:
+            if use_transposed:
+                transposed_circ = decomposition.copy_empty_like()
+                # QuantumCircuits get a default name, while DAGCircuits don't
+                transposed_circ.name = decomposition.name or "synth_circuit"
+                transposed_circ.name += "_transpose"
+                for node in decomposition.topological_op_nodes():
+                    if node.op.name != "cx":
+                        raise CircuitError("The circuit contains non-CX gates.")
+                    transposed_circ.apply_operation_front(
+                        node.op, reversed(node.qargs), node.cargs, check=False
+                    )
+                decomposition = transposed_circ
+            if use_inverted:
+                inverted_circ = decomposition.copy_empty_like()
+                for node in decomposition.topological_op_nodes():
+                    inverted_circ.apply_operation_front(
+                        node.op.inverse(), node.qargs, node.cargs, check=False
+                    )
+                decomposition = inverted_circ
+            return decomposition
```

Comment:
This is another place where (imho) supporting both `QuantumCircuit` and `DAGCircuit` makes the code quite a bit more complex. I am wondering how much performance improvement we get in practice.